File: Multidimensional/multidimensional-v2.py
#import matplotlib.pyplot as plt
import numpy as np

#################################################### INICIALIZACAO ##############################################

#Constantes
sigma = 20
deltaT = 0.25                                               #Variacao de tempo
epsilon = 0.1
gravidade = 9.8
aceleracao = 30

#Predicao da posicao e tempo: x = Fx + Gu
x = np.zeros((2, 1))                                        #Inicializacao da matriz x 2x1 de estimativas

F = np.zeros((2, 2))                                        #Inicializacao da matriz F 2x2
for j in range(0, 2):
    F[j, j] = 1                                             #Preenche diagonal principal com 1
F[0, 1] = deltaT

G = np.zeros((2, 1))                                        #Inicializacao da matriz G 1x2
G[0, 0] = 0.5*(deltaT**2)
G[1, 0] = deltaT

#u = aceleracao + gravidade                                  #Inicializacao da matriz u 1x1 de controle
u = 0

#Predicao da covariancia: P = FP(F^T)+Q
P = np.zeros((2, 2))                                        #Inicializacao de uma covariancia arbitraria
for j in range(0, 2):
    P[j, j] = 500

Q = np.zeros((2, 2))                                        #Inicializacao da matriz Q 2x2
Q[0, 0] = (deltaT**4)/4
Q[0, 1] = Q[1, 0] = (deltaT**3)/2
Q[1, 1] = deltaT**2
Q = Q*(epsilon**2)

R = [sigma**2]                                              #Inicializacao do vetor R

H = np.zeros((1, 2))                                         #Inicializacao da matriz H
H[0, 0] = 1

I = np.zeros((2, 2))                                        #Inicializacao da matriz identidade I 2x2
for j in range(0, 2):
    I[j, j] = 1

Ft = np.transpose(F)                                        #Inicializacao da matriz transposta de F
Ht = np.transpose(H)                                        #Inicializacao da matriz transposta de H

######################################## FILTRO DE KALMAN MULTIDIMENSIONAL ######################################

for j in range(0, 30):
    x = F @ x + G * u  # Predicao da estimativa
    P = F @ P @ Ft + Q                                          #Predicao da covariancia

    K = P @ Ht @ np.linalg.inv(H @ P @ Ht + R)                  #Predicao do Ganho de Kalman
    Kt = np.transpose(K)                                        #Matriz transposta de K

    z = float(input('Medicao da posicao z: '))                  #Input do usuario
    u = float(input('Medicao do controle u: '))                 #Input do usuario

    u = u - gravidade                                           #Correcao do controle, devido ao acelerometro

    x = x + K @ (z - H @ x)                                         #Atualizacao da matriz de estimativas
    P = (I - K @ H) @ P @ (np.transpose(I - K @ H)) + K * R @ Kt    #Atualizacao da matriz de covariancias

File: Multidimensional/multidimensional-v2.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% DADOS INICIAIS %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%Constantes
sigma = 20;
tempo = 0.25;
epsilon = 0.1;
gravidade = 9.8;
aceleracao = 30;

%Equacao de predicao: x = Fx+Gu
x = zeros(2, 1);                                        %Inicializacao da matriz 2x1 de estimativa
F = [1 tempo ; 0 1];                                    %Inicializacao da matriz F
G = [0.5*tempo^2 ; tempo];                              %Inicializacao da matriz G
u = 0;                                                  %Inicializacao da variavel de input

%Equacao de predicao da covariancia: P = F*P*Ft+Q
P = [500 0 ; 0 500];                                    %Inicializacao de uma covariancia arbitraria
Q = [(tempo^4)/4 (tempo^3)/2 ; (tempo^3)/2 tempo^2]*epsilon^2;

H = eye(2)*1.5;

R = sigma^2;

I = eye(2);                                             %Matriz identidade 6x6

Ft = transpose(F);
Ht = transpose(H);

z = zeros(2, 1);                                        %Inicializacao da matriz 1x2 de medicao. z = [altura aceleracao]

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

estimados = zeros(2,10);                               %Inicializacao da matriz de estimativa, com x dados em i tempos
medidos = zeros(2,10);                                 %Inicializacao da matriz auxiliar de medicao, com x dados em i tempos

%Laco de repeticao para medicao e estimacao de i tempos
for i = 1:10

    x = F*x + G*u;                                      %Predicao da estimativa 
    P = F*P*Ft + Q;                                     %Predicao da covariancia
    K = P*Ht*(H*P*Ht + R)^(-1);                         %Predicao do ganho de Kalman

    %Entrada do usuario   
    z(1) = input(sprintf('Digite o valor da medicao z: '));         %Medicao da altura
    z(2) = input(sprintf('Digite o valor do input u: '));           %Medicao do controle

    z(2) = z(2) - gravidade;

    Kt = transpose(K);                                  %Atualizacao da matriz transposta do ganho de Kalman                                  
    x = x + K*(z - H*x);
    P = (I - K*H)*P*(transpose(I - K*H)) + K*R*Kt;      %Atualizacao da matriz de covariancia

    estimados(:,i) = x;                                 %Matriz de estimados recebe x, no tempo i
    medidos(:,i) = z;                                   %Matriz de medidos recebe z, no tempo i                               

end

estimado_altura = estimados(1,:);                              %Armazena as estimativas num vetor
estimado_velocidade = estimados(2,:);                          %Armazena as velocidadades num vetor
medido_altura = medidos(1,:);
medido_velocidade= medidos(2,:);

%Plot do grafico de altura
figure
plot(estimado_altura,'b')
hold on
plot(medido_altura,'r')
xlabel('Tempo (s)');
ylabel('Altura (m)');
title('Estimativa vs Medicao da altura');

%Plot do grafico de velocidade
figure
plot(estimado_velocidade,'b')
hold on
plot(medido_velocidade,'r')
xlabel('Tempo (s)');
ylabel('Velocidade (m)');
title('Estimativa vs Medicao da velocidade');
